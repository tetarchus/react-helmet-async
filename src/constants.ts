const TAG_PROPERTIES = {
  CHARSET: 'charset',
  CSS_TEXT: 'cssText',
  HREF: 'href',
  HTTPEQUIV: 'http-equiv',
  INNER_HTML: 'innerHTML',
  ITEM_PROP: 'itemprop',
  NAME: 'name',
  PROPERTY: 'property',
  REL: 'rel',
  SRC: 'src',
};

const ATTRIBUTE_NAMES = {
  BODY: 'bodyAttributes',
  HTML: 'htmlAttributes',
  TITLE: 'titleAttributes',
};

const TAG_NAMES = {
  BASE: 'base',
  BODY: 'body',
  HEAD: 'head',
  HTML: 'html',
  LINK: 'link',
  META: 'meta',
  NOSCRIPT: 'noscript',
  SCRIPT: 'script',
  STYLE: 'style',
  TITLE: 'title',
  FRAGMENT: 'Symbol(react.fragment)',
};

const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

const VALID_TAG_NAMES = Object.keys(TAG_NAMES).map((tagName) => TAG_NAMES[tagName] as string);

const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

const HTML_TAG_MAP = Object.fromEntries(
  Object.keys(REACT_TAG_MAP).map((key) => [REACT_TAG_MAP[key as keyof typeof REACT_TAG_MAP], key]),
);

const HELMET_ATTRIBUTE = 'data-rh';

const HELMET_PROPS = {
  DEFAULT_TITLE: 'defaultTitle',
  DEFER: 'defer',
  ENCODE_SPECIAL_CHARACTERS: 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE: 'onChangeClientState',
  TITLE_TEMPLATE: 'titleTemplate',
  PRIORITIZE_SEO_TAGS: 'prioritizeSeoTags',
};

export {
  ATTRIBUTE_NAMES,
  HELMET_ATTRIBUTE,
  HELMET_PROPS,
  HTML_TAG_MAP,
  REACT_TAG_MAP,
  SEO_PRIORITY_TAGS,
  TAG_NAMES,
  TAG_PROPERTIES,
  VALID_TAG_NAMES,
};
