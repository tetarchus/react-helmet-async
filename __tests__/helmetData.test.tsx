import React from 'react';

import { Helmet, HelmetData } from '..';
import { HELMET_ATTRIBUTE } from '../components/constants';

import { after, before, render } from './setup/testSetup';

describe('Helmet Data', () => {
  beforeEach(before);
  afterEach(after);

  describe('Browser', () => {
    it('renders without context', () => {
      expect.assertions(6);

      const helmetData = new HelmetData({});

      render(
        <Helmet
          base={{ target: '_blank', href: 'http://localhost/' }}
          defer={false}
          helmetData={helmetData}
        />,
      );

      const existingTags = document.head.querySelectorAll(`base[${HELMET_ATTRIBUTE}]`);
      const [firstTag] = Array.prototype.slice.call(existingTags) as Element[];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag.getAttribute('href')).toBe('http://localhost/');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('renders declarative without context', () => {
      expect.assertions(6);

      const helmetData = new HelmetData({});

      render(
        <Helmet defer={false} helmetData={helmetData}>
          <base target='_blank' href='http://localhost/' />
        </Helmet>,
      );

      const existingTags = document.head.querySelectorAll(`base[${HELMET_ATTRIBUTE}]`);
      const [firstTag] = Array.prototype.slice.call(existingTags) as Element[];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag.getAttribute('href')).toBe('http://localhost/');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('sets base tag based on deepest nested component', () => {
      expect.assertions(6);

      const helmetData = new HelmetData({});

      render(
        <div>
          <Helmet defer={false} helmetData={helmetData}>
            <base href='http://mysite.com' />
          </Helmet>
          <Helmet defer={false} helmetData={helmetData}>
            <base href='http://mysite.com/public' />
          </Helmet>
        </div>,
      );

      const existingTags = document.head.querySelectorAll(`base[${HELMET_ATTRIBUTE}]`);
      const [firstTag] = Array.prototype.slice.call(existingTags) as Element[];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag.getAttribute('href')).toBe('http://mysite.com/public');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });
  });

  describe('Server', () => {
    it('renders without context', () => {
      expect.assertions(3);

      const helmetData = new HelmetData({});

      render(
        <Helmet
          base={{ target: '_blank', href: 'http://localhost/' }}
          defer={false}
          helmetData={helmetData}
        />,
        true,
      );

      const head = helmetData.context.helmet;

      expect(head?.base).toBeDefined();
      expect(head?.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });

    it('renders declarative without context', () => {
      expect.assertions(3);

      const helmetData = new HelmetData({});

      render(
        <Helmet defer={false} helmetData={helmetData}>
          <base target='_blank' href='http://localhost/' />
        </Helmet>,
        true,
      );

      const head = helmetData.context.helmet;

      expect(head?.base).toBeDefined();
      expect(head?.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });

    it('sets base tag based on deepest nested component', () => {
      expect.assertions(3);

      const helmetData = new HelmetData({});

      render(
        <div>
          <Helmet defer={false} helmetData={helmetData}>
            <base href='http://mysite.com' />
          </Helmet>
          <Helmet defer={false} helmetData={helmetData}>
            <base href='http://mysite.com/public' />
          </Helmet>
        </div>,
        true,
      );

      const head = helmetData.context.helmet;

      expect(head?.base).toBeDefined();
      expect(head?.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });

    it('works with the same context object but separate HelmetData instances', () => {
      expect.assertions(3);

      const initialContext = {};

      const { context } = render(
        <div>
          <Helmet defer={false} helmetData={new HelmetData(initialContext)}>
            <base href='http://mysite.com' />
          </Helmet>
          <Helmet defer={false} helmetData={new HelmetData(initialContext)}>
            <base href='http://mysite.com/public' />
          </Helmet>
        </div>,
        true,
      );

      const head = 'helmet' in context ? context.helmet : null;

      expect(head?.base).toBeDefined();
      expect(head?.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });
  });
});
