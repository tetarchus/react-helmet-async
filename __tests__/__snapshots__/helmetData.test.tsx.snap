// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helmet Data Browser renders declarative without context 1`] = `"<base target=\\"_blank\\" href=\\"http://localhost/\\" data-rh=\\"true\\">"`;

exports[`Helmet Data Browser renders without context 1`] = `"<base target=\\"_blank\\" href=\\"http://localhost/\\" data-rh=\\"true\\">"`;

exports[`Helmet Data Browser sets base tag based on deepest nested component 1`] = `"<base href=\\"http://mysite.com/public\\" data-rh=\\"true\\">"`;

exports[`Helmet Data Server renders declarative without context 1`] = `"<base data-rh=\\"true\\" target=\\"_blank\\" href=\\"http://localhost/\\"></base>"`;

exports[`Helmet Data Server renders without context 1`] = `"<base data-rh=\\"true\\" target=\\"_blank\\" href=\\"http://localhost/\\"></base>"`;

exports[`Helmet Data Server sets base tag based on deepest nested component 1`] = `"<base data-rh=\\"true\\" href=\\"http://mysite.com/public\\"></base>"`;

exports[`Helmet Data Server works with the same context object but separate HelmetData instances 1`] = `""`;
